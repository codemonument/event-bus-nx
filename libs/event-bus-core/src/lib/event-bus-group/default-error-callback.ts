export function defaultErrorCallback(error: unknown) {
  console.error(
    `[event-bus webcomponent]: An Error was caught while calling a listener callback.
  This is the default ErrorCallbackFunction, which can be replaced by calling setDefaultErrorCallback() on the event-bus dom node,
  or passing the optional 'errorCallback' parameter when calling on(EventType, ...) function.

  Error Details:

  `,
    error,
  );
}
