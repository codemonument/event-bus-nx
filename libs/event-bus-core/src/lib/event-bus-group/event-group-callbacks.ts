import { EventualPayload } from "../types/bus-event.type";

/**
 * Is called 'CallbackFunction' in original article
 */
export type EventGroupCallback<P> = (eventPayload: EventualPayload<P>) => void;

/**
 * Is called 'CallbackFunction' in original article
 */
export type EventGroupErrorCallback<T> = (error: T) => void;

export function defaultErrorCallback(error: unknown) {
  console.error(
    `[event-bus webcomponent]: An Error was caught while calling a listener callback.
  This is the default ErrorCallbackFunction, which can be replaced by calling setDefaultErrorCallback() on the event-bus dom node,
  or passing the optional 'errorCallback' parameter when calling on(EventType, ...) function.

  Error Details:

  `,
    error,
  );
}
