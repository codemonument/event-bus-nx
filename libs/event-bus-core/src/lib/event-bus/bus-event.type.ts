/**
 * When P = void, resulting type MUST be void, otherwise typescript forces the user
 * to input a payload param when instantiating a child class of BusEvent
 */
export type EventualPayload<P> = P extends void ? void : P;

/**
 * I am the base-class for all of the events that this application pushes onto the
 * EventBus. The only guarantee that this class makes is a read-only property called `type`.
 *
 * This class has been called 'Event' in original Article.
 *
 * It also allows for a payload being passed into into it.
 * Can be disabled when not needed by simply extending like this with `void`:
 *
 * class EventWithoutPayload extends BusEvent<void> {
 *   public type = "EventWithoutPayload";
 * }
 *
 * @type G_PayloadType
 * === generic payload type
 * === "type variable" for the typescript type of the payload incomming into the constructor.
 */
export abstract class BusEvent<G_PayloadType> {
  public abstract readonly type: string;
  constructor(public readonly payload: EventualPayload<G_PayloadType>) {}
}

/**
 * Allows to ensure, that a certain type
 * - can be instantiated with calling `new`
 *   (this is needed to be able to use the type in question with `instanceof`)
 * - has a parameter `payload` which is of type P, unles
 * - P is `void`, in this case it's not allowed to pass anything to this constructor!
 *   (expressed via the conditional `never` type)
 *
 * Technical Details:
 * Actually, this interface does NOT define the class type,
 * but defines the shape of a FUNCTION (like the @FunctionalInterface in Java).
 * Reason: A typescript class name (like Map) internally refers to the constructor of a class,
 * which happens to have more things attached to it's prototype!
 *
 * Imagination:
 * Think of NewableBusEvent interface as a Factory, which nows how to construct stuff with new.
 * In this case, it knows, how to construct instances of E (the event type)
 * See: https://stackoverflow.com/questions/13407036/how-does-interfaces-with-construct-signatures-work
 */
export interface NewableBusEvent<E extends BusEvent<P>, P> {
  // P - Define that the constructor gets a Payload of type EventualPayload<P>
  new <P>(payload: EventualPayload<P>): E;
}
