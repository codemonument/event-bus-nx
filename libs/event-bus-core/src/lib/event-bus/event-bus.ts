import { filter, map, Observable, Subject } from "rxjs";
import {
  BusEvent,
  EventualPayload,
  NewableBusEvent,
} from "./types/bus-event.type";
import { SimpleNewable } from "./types/simple-newable.type";

export class EventBus {
  private eventStream: Subject<unknown> = new Subject<unknown>();

  // --- PUBLIC FUNCTIONS ---

  /**
   * I filter the event stream to get only events of one type as observable
   * @param typeFilter The event type to listen to
   *        must extend from BusEvent<R>
   *
   * @returns either the Event E or the Payload of E, typed P
   * Note: These types do not be passed manually, they will be inferenced by TS
   */
  public on$<E extends BusEvent<P>, P>(
    typeFilter: NewableBusEvent<E, P>,
  ): Observable<EventualPayload<P>> {
    return this.eventStream.pipe(
      // Filters all events on the event stream and returns only these, which map the typeFilter
      filter((event: unknown): event is E => {
        return event instanceof typeFilter;
      }),
      // Maps the events to their payloads for easier consumption
      // Note: The return type must be EventualPayload here
      //       to not get `P | undefined` as return type of this map
      map((event): EventualPayload<P> =>
        (event.payload !== undefined)
          ? event.payload
          : undefined as EventualPayload<P>
      ),
    );
  }

  /**
   * I push the given event onto the message bus.
   *
   * @param event
   */
  public emit(event: unknown): void {
    this.eventStream.next(event);
  }

  /**
   * CAUTION: PLEASE BE SURE WHAT YOU DO WHEN YOU USE THIS!
   *
   * Returns the whole eventStream as observable.
   */
  public get eventStream$() {
    return this.eventStream.asObservable();
  }
}
