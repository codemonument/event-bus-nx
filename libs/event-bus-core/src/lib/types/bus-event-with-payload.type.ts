import { BusEvent } from "./bus-event.type";

/**
 * I am the sub-class / base-class for all of the payload-heavy events that this
 * application pushes onto the MessageQueue. This class guarantees a payload with a
 * given interface.
 *
 * Was called 'EventWithPayload' in original article
 */
export abstract class BusEventWithPayload<T> extends BusEvent {
  public readonly payload: T;

  constructor(payload: T) {
    super();
    this.payload = payload;
  }
}
