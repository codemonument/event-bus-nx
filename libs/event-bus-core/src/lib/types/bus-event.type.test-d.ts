// Readme about syntax: https://www.npmjs.com/package/tsd

// import { expectType } from "tsd";
// import concat from ".";

// expectType<string>(concat("foo", "bar"));
// expectType<string>(concat(1, 2));
