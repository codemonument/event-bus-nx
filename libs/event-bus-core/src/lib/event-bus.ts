import { filter, Observable, Subject } from "rxjs";
import { TypeDiscriminator } from "./types/type-discriminator";

export class EventBus {
  private eventStream: Subject<unknown> = new Subject<unknown>();

  // --- PUBLIC FUNCTIONS ---

  /**
   * I filter the event stream to get only events of one type as observable
   * @param typeFilter The event type to listen to
   * @returns
   */
  public on$<T>(typeFilter: TypeDiscriminator<T>): Observable<T> {
    return this.eventStream.pipe(
      filter((event: unknown): event is T => {
        return event instanceof typeFilter;
      }),
    );
  }

  /**
   * I push the given event onto the message bus.
   *
   * @param event
   */
  public emit(event: unknown): void {
    this.eventStream.next(event);
  }

  /**
   * PLEASE BE SURE WHAT YOU DO WHEN YOU USE THIS!
   * Return the whole eventStream as observable
   */
  public get eventStream$() {
    return this.eventStream.asObservable();
  }
}
