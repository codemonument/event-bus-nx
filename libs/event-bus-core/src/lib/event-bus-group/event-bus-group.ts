import { Subscription } from "rxjs";
import { EventBus } from "../event-bus/event-bus";
import {
  BusEvent,
  EventualPayload,
  NewableBusEvent,
  payloadOf,
} from "../event-bus/bus-event.type";
import {
  defaultErrorCallback,
  EventGroupCallback,
  EventGroupErrorCallback,
} from "./event-group-callbacks.type";

/**
 * An EventBusGroup allows for using the event bus with a Callback-Interface.
 * The this.on() method allows to register a callback to a certain event type.
 * The rxjs subscription on the EventBus, which is made during the callback registration,
 * will be stored inside the EventBusGroup.
 * This allows for unsubscribing all callback subscriptions at once
 */
export class EventBusGroup {
  private subscriptions: Subscription[] = [];

  constructor(
    private bus: EventBus,
    private errorCallback: EventGroupErrorCallback = defaultErrorCallback,
  ) {
  }

  /**
   * Replaces the default error callback function with a custom one
   */
  public setDefaultErrorCallback(callback: EventGroupErrorCallback): void {
    this.errorCallback = callback;
  }

  /**
   * I subscribe to the message bus, but only invoke the callback when the event is
   * of the given newable type (ie, it's a Class definition, not an instance).
   * --
   * NOTE: The NewableBusEvent<E, P> enables internal type inference.
   *
   * @param typeFilter
   * @param callback
   * @param callbackContext
   * @returns
   */
  public on<E extends BusEvent<payloadOf<E>>>(
    typeFilter: NewableBusEvent<E>,
    callback: EventGroupCallback<payloadOf<E>>,
    callbackContext: unknown = null,
    errorCallback?: EventGroupErrorCallback,
  ): void {
    const next = (eventPayload: EventualPayload<payloadOf<E>>) => {
      try {
        callback.call(callbackContext, eventPayload);
      } catch (error) {
        if (errorCallback) {
          errorCallback.call(callbackContext, errorCallback);
          return;
        }

        this.errorCallback.call(callbackContext, error);
      }
    };

    const sub = this.bus.on$(typeFilter).subscribe(
      {
        next,
        error: (error: unknown) =>
          this.errorCallback.call(callbackContext, error),
      },
    );

    this.subscriptions.push(sub);
  }

  public unsubscribeAll() {
    for (const sub of this.subscriptions) {
      sub.unsubscribe();
    }
  }
}
