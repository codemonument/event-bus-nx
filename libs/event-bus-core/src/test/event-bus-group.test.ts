import { EventBus } from "../lib/event-bus/event-bus";
import { EventBusGroup } from "../lib/event-bus-group/event-bus-group";
import { EventWithPayload, PlainEvent } from "./test-events.types";

describe(`event-bus-group.test`, () => {
  it(`should be constructed`, () => {
    const eBus = new EventBus();
    const eGroup = new EventBusGroup(eBus);
    expect(eGroup).toBeDefined();
  });

  it(`should receive PlainEvent in callback (test event registration)`, (done) => {
    const eBus = new EventBus();
    const eGroup = new EventBusGroup(eBus);

    eGroup.on(PlainEvent, () => {
      done();
    });

    eBus.emit(new PlainEvent());
  });

  it(`should receive DemoPayload in callback (test EventWithPayload)`, (done) => {
    const eBus = new EventBus();
    const eGroup = new EventBusGroup(eBus);
    const demoEvent = new EventWithPayload({ name: "Bob" });

    eGroup.on(EventWithPayload, ({ name }) => {
      expect(name).toEqual(demoEvent.payload.name);
      done();
    });

    eBus.emit(demoEvent);
  });

  it(`should unsubscribe all event callbacks`, () => {
    const eBus = new EventBus();
    const eGroup = new EventBusGroup(eBus);

    eGroup.on(PlainEvent, () => {
      console.log("Received Plain Event");
    });

    eGroup.on(EventWithPayload, (payload) => {
      console.log("Received Event with Payload", payload);
    });

    // this should not throw - test fails if it does throw
    eGroup.unsubscribeAll();
  });

  // TODO: Finish EventBusGroup Tests!
});
