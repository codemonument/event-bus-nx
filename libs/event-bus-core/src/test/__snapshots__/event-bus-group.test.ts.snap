// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`event-bus-group.test should trigger custom error callback from EventBusGroup constructor 1`] = `[Error: Fake error in callback!]`;

exports[`event-bus-group.test should trigger custom error callback from eGroup.on() 1`] = `[Error: Fake error in callback!]`;
